src/Workspaces/Remote/ServiceHub/Services/Renamer/RemoteRenamerService.cs

```diff
@@ -6,26 +6,56 @@
 using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis.Rename;
 using Microsoft.CodeAnalysis.Rename.ConflictEngine;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Remote
 {
     internal sealed class RemoteRenamerService : BrokeredServiceBase, IRemoteRenamerService
     {
-        internal sealed class Factory : FactoryBase<IRemoteRenamerService>
+        internal sealed class Factory : FactoryBase<IRemoteRenamerService, IRemoteRenamerService.ICallback>
         {
-            protected override IRemoteRenamerService CreateService(in ServiceConstructionArguments arguments)
-                => new RemoteRenamerService(arguments);
+            protected override IRemoteRenamerService CreateService(in ServiceConstructionArguments arguments, RemoteCallback<IRemoteRenamerService.ICallback> callback)
+                => new RemoteRenamerService(arguments, callback);
         }
 
-        public RemoteRenamerService(in ServiceConstructionArguments arguments)
+        private readonly RemoteCallback<IRemoteRenamerService.ICallback> _callback;
+
+        public RemoteRenamerService(in ServiceConstructionArguments arguments, RemoteCallback<IRemoteRenamerService.ICallback> callback)
             : base(arguments)
         {
+            _callback = callback;
         }
 
+#if TODO // Replace the below specialization with a call to a generic impl once https://github.com/microsoft/vs-streamjsonrpc/issues/789 is fixed
```

Comment:
Creates options provider that fetches options from the client and caches them for a given language.